src/appleseed/renderer/modeling/postprocessingstage/vignettepostprocessingstage.cpp

```diff
@@ -0,0 +1,198 @@
+
+    const float DefaultIntensity = 0.5f;
+    const float MinIntensity = 0.0f;
+    const float MaxIntensity = 1.0f;
+
+    class VignettePostProcessingStage
+        : public PostProcessingStage
+    {
+      public:
+        VignettePostProcessingStage(
+            const char*             name,
+            const ParamArray&       params)
+          : PostProcessingStage(name, params)
+        {
+        }
+
+        void release() override
+        {
+            delete this;
+        }
+
+        const char* get_model() const override {
+            return Model;
+        }
+
+        bool on_frame_begin(
+            const Project&          project,
+            const BaseGroup*        parent,
+            OnFrameBeginRecorder&   recorder,
+            IAbortSwitch*           abort_switch) override
+        {
+            const OnFrameBeginMessageContext context("post-processing stage", this);
+
+            m_intensity = m_params.get_optional("intensity", DefaultIntensity, context);
+
+            return true;
+        }
+
+        void execute(Frame& frame) const override
+        {
+            const CanvasProperties& props = frame.image().properties();
+            const Vector2f resolution(static_cast<float>(props.m_canvas_width), static_cast<float>(props.m_canvas_height));
+            const Vector2f aspect(resolution.x / resolution.y, 1);
+
+            Image& image = frame.image();
+
+            for (size_t y = 0; y < resolution.y; ++y)
+            {
+                for (size_t x = 0; x < resolution.x; ++x)
+                {
+                    Vector2f coord = (2.0f * Vector2f(static_cast<float>(x), static_cast<float>(y)) - resolution) / resolution.y;
+
+                    //
+                    // Port of Keijiro Takahashi's natural vignetting effect for Unity.
+                    //
+                    // Reference:
+                    //
+                    //   https://github.com/keijiro/KinoVignette
+                    //   https://en.wikipedia.org/wiki/Vignetting#Natural_vignetting
+                    //
+
+                    // Recreates natural illumination falloff, which is approximated by the "cosine fourth" law of illumination falloff.
+                    float rf = norm(coord) * m_intensity; // radial falloff
+                    float rf2_1 = rf * rf + 1.0f;
+                    float e = 1.0f / (rf2_1 * rf2_1); // inversely proportional to the fourth power of the distance to the center
+
+                    Color4f background_premult;
+                    image.get_pixel(x, y, background_premult);
+
+                    background_premult.r *= e;
+                    background_premult.g *= e;
+                    background_premult.b *= e;
+
+                    image.set_pixel(
+                        x, y,
+                        background_premult);
+                }
+            }
+        }
+
+      private:
+          float m_intensity;
```

Comment:
Indent just once (4 spaces) with respect to the class. Basically only `public`, `private`, etc. are half-indented, everything else is indented as usual.